tests/Unit/DataStructures/Tensor/Expressions/Test_Product.cpp

```diff
@@ -0,0 +1,1099 @@
+
+  Tensor<DataType, Symmetry<1>,
+         index_list<SpacetimeIndex<3, UpLo::Up, Frame::Inertial>>>
+      Su(used_for_size);
+  create_tensor(make_not_null(&Su));
+
+  // \f$L^{a} = R * S^{a}\f$
+  // Use explicit type (vs auto) for LHS Tensor so the compiler checks the
+  // return type of `evaluate`
+  const decltype(Su) LA_from_R_SA =
+      TensorExpressions::evaluate<ti_A>(R() * Su(ti_A));
+  // \f$L^{a} = S^{a} * R\f$
+  const decltype(Su) LA_from_SA_R =
+      TensorExpressions::evaluate<ti_A>(Su(ti_A) * R());
+
+  for (size_t a = 0; a < 4; a++) {
+    CHECK(LA_from_R_SA.get(a) == R.get() * Su.get(a));
+    CHECK(LA_from_SA_R.get(a) == Su.get(a) * R.get());
+  }
+
+  Tensor<DataType, Symmetry<2, 1>,
+         index_list<SpacetimeIndex<3, UpLo::Lo, Frame::Inertial>,
+                    SpatialIndex<4, UpLo::Lo, Frame::Inertial>>>
+      Tll(used_for_size);
+  create_tensor(make_not_null(&Tll));
+
+  // \f$L_{ai} = R * T_{ai}\f$
+  const Tensor<DataType, Symmetry<2, 1>,
+               index_list<SpacetimeIndex<3, UpLo::Lo, Frame::Inertial>,
+                          SpatialIndex<4, UpLo::Lo, Frame::Inertial>>>
+      Lai_from_R_Tai =
+          TensorExpressions::evaluate<ti_a, ti_i>(R() * Tll(ti_a, ti_i));
+  // \f$L_{ia} = R * T_{ai}\f$
+  const Tensor<DataType, Symmetry<2, 1>,
+               index_list<SpatialIndex<4, UpLo::Lo, Frame::Inertial>,
+                          SpacetimeIndex<3, UpLo::Lo, Frame::Inertial>>>
+      Lia_from_R_Tai =
+          TensorExpressions::evaluate<ti_i, ti_a>(R() * Tll(ti_a, ti_i));
+  // \f$L_{ai} = T_{ai} * R\f$
+  const Tensor<DataType, Symmetry<2, 1>,
+               index_list<SpacetimeIndex<3, UpLo::Lo, Frame::Inertial>,
+                          SpatialIndex<4, UpLo::Lo, Frame::Inertial>>>
+      Lai_from_Tai_R =
+          TensorExpressions::evaluate<ti_a, ti_i>(Tll(ti_a, ti_i) * R());
+  // \f$L_{ia} = T_{ai} * R\f$
+  const Tensor<DataType, Symmetry<2, 1>,
+               index_list<SpatialIndex<4, UpLo::Lo, Frame::Inertial>,
+                          SpacetimeIndex<3, UpLo::Lo, Frame::Inertial>>>
+      Lia_from_Tai_R =
+          TensorExpressions::evaluate<ti_i, ti_a>(Tll(ti_a, ti_i) * R());
+
+  for (size_t a = 0; a < 4; a++) {
+    for (size_t i = 0; i < 4; i++) {
+      const DataType expected_R_Tai_product = R.get() * Tll.get(a, i);
+      CHECK(Lai_from_R_Tai.get(a, i) == expected_R_Tai_product);
+      CHECK(Lia_from_R_Tai.get(i, a) == expected_R_Tai_product);
+
+      const DataType expected_Tai_R_product = Tll.get(a, i) * R.get();
+      CHECK(Lai_from_Tai_R.get(a, i) == expected_Tai_R_product);
+      CHECK(Lia_from_Tai_R.get(i, a) == expected_Tai_R_product);
+    }
+  }
+}
+
+/// \ingroup TestingFrameworkGroup
+/// \brief Test the outer product of rank 1 tensors with another tensor is
+/// correctly evaluated
+///
+/// \details
+/// The outer product cases tested are:
+/// - \f$L^{a}{}_{i} = R_{i} * S^{a}\f$
+/// - \f$L_{ja}{}^{i} = R_{i} * S^{a} * T^{j}\f$
```

Comment:
The index up/down-ness doesn't match on LHS and RHS here

Reply:
Good catch!